_posts/2019-09-25-scatter-gather-pt1.md

```diff
@@ -0,0 +1,99 @@
+{
+    foreach (var orderItem in message.OrderItems)
+    {
+        await context.Send(new ProcessOrderItem
+        {
+            OrderId = message.OrderId,
+            ItemId = orderItem.Id
+        });
+    }
+}
+```
+
+Based on an incoming message, the snippet above splits the items in the order into separate processing commands and sends them out for parallel processing. NServiceBus takes care of batching outgoing messages: If the handler fails with an exception at any point, no message is sent.
+
+### Aggregate
+
+Whether we use events or commands, in both cases we expect multiple responses and we want to consolidate their results. This means we need some persistent state across all responses which stores each response's result. Very often, people decide to use a [Saga](https://docs.particular.net/nservicebus/sagas) for this, as it provides a shared state across multiple message handlers out of the box. The Scatter-Gather implementation would probably look something like this:
+
+```
+class ScatterGather : Saga<ScatterGatherSagaData>,
+    IAmStartedByMessages<OrderPlaced>,
+    IHandleMessages<ProcessOrderItemResult>
+{
+    public async Task Handle(OrderPlaced message, IMessageHandlerContext context)
+    {
+        Data.NumberOfItems = message.OrderItems.Count();
+        foreach (var orderItem in message.OrderItems)
+        {
+            await context.Send(new ProcessOrderItem
+            {
+                OrderId = message.OrderId,
+                ItemId = orderItem.Id
+            });
+        }
+    }
+
+    public async Task Handle(ProcessOrderItemResult message, IMessageHandlerContext context)
+    {
+        var totalProcessedOrders = ++Data.ProcessedOrders;
+        if (totalProcessedOrders == Data.NumberOfItems)
+        {
+            await context.Publish(new OrderProcessed { OrderId = Data.OrderId });
+            MarkAsComplete();
+            Console.WriteLine("Order completed");
+        }
+    }
+    
+    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<ScatterGatherSagaData> mapper)
+    {
+        mapper.ConfigureMapping<OrderPlaced>(m => m.OrderId).ToSaga(s => s.OrderId);
+        mapper.ConfigureMapping<ProcessOrderItemResult>(m => m.OrderId).ToSaga(s => s.OrderId);
+    }
+}
+```
+
+This looks quite nice, the Saga API provides a simple solution to implement the Scatter-Gather pattern. Unfortunately, it's a very suboptimal solution that will run into issues at runtime. 
+
+### Sagas and concurrency
+
+When a handler of a saga is invoked, the state of the saga is loaded from the database. This leads to a situation where multiple responses want to update the same saga instance concurrently (keep in mind that NServiceBus can handle multiple messages concurrently by default). Sagas can't handle concurrency, there is only one message which can successfully complete a saga at one time because they all touch the same data.
```

Comment:
“Can’t handle” reads as a limitation. As something that cannot be done. I guess that what you want to say is that sagas are designed o handle concurrency in such a way that it doesn’t result in an inconsistent state.